### PR TITLE
CEDS-1595 Associate MUCR 

### PR DESCRIPTION
**Pull request by bambuchaAdm**

### CODE REVIEW

File: app/forms/AssociateUcr.scala [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+      data.get(key).map { kind =>
+        lookup.andThen(Right.apply).applyOrElse(kind, (_: String) => Left(Seq(FormError(key, "error.unknown"))))
+      }
+    }.getOrElse(Left(Seq(FormError(key, "error.required"))))
+
+    override def unbind(key: String, value: AssociateKind): Map[String, String] = Map(key -> value.formValue)
+  }
+
+  implicit val format =
+    Format[AssociateKind](Reads.StringReads.collect(JsonValidationError("error.unknown"))(lookup), Writes(kind => JsString(kind.formValue)))
```
Comment (ed0906) [inline app/forms/AssociateUcr.scala]: What's wrong with Json.format?

File: app/forms/AssociateUcr.scala [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+object AssociateKind {
+  case object Mucr extends AssociateKind("mucr")
+  case object Ducr extends AssociateKind("ducr")
+
+  private val lookup = PartialFunction[String, AssociateKind] {
+    case Mucr.formValue => Mucr
+    case Ducr.formValue => Ducr
+  }
+
+  implicit val formatter: Formatter[AssociateKind] = new Formatter[AssociateKind] {
```
Comment (ed0906) [inline app/forms/AssociateUcr.scala]: Whats this for?

File: test/views/spec/ViewMatchers.scala [hunk 3]
```diff
@@ -216,4 +217,11 @@ trait ViewMatchers {
   def haveHref(url: String): Matcher[Element] = new ElementHasAttributeValueMatcher("href", url)
 
   def haveTranslationFor(key: String): Matcher[Messages] = new TranslationKeyMatcher(key)
+
+  val checked: BeMatcher[Element] = new BeMatcher[Element] {
+    override def apply(left: Element): MatchResult =
+      MatchResult(left.attr("checked") == "checked", "Element is not checked", "Element is checked")
+  }
+
+  val unchecked: BeMatcher[Element] = not(checked)
```
Comment (ed0906) [inline test/views/spec/ViewMatchers.scala]: 👍

File: app/views/associate_ucr.scala.html [hunk 4]
```diff
@@ -14,29 +14,43 @@
...
+
+        @components.input_radio(
+            field = form("kind"),
+            legend = "",
+            hint = Some(messages("associate.ucr.hint")),
+            inputs = Seq(
+                RadioOption(
+                    keyPrefix = "associate.ucr",
+                    option = AssociateKind.Ducr.formValue,
+                    revealComponent = components.input_text(form("ducr"), "") // FIXME accessability
```
Comment (ed0906) [inline app/views/associate_ucr.scala.html]: Are these getting fixed as part of this task?

File: app/views/associate_ucr.scala.html [hunk 5]
```diff
@@ -14,29 +14,43 @@
...
+                ),
+                RadioOption(
+                    keyPrefix = "associate.ucr",
+                    option = AssociateKind.Mucr.formValue,
+                    revealComponent = components.input_text(form("mucr"), "") // FIXME accessability
```
Comment (ed0906) [inline app/views/associate_ucr.scala.html]: Are these getting fixed as part of this task?

File: app/forms/AssociateUcr.scala (hunk 2 above)
Comment (bambuchaAdm) [reply app/forms/AssociateUcr.scala]: Mapping

Comment (ed0906) [reply app/forms/AssociateUcr.scala]: Mapping what?

Comment (bambuchaAdm) [reply app/forms/AssociateUcr.scala]: without `Formatter` i am unable to use this class as part of mapping of `AssociateUcr`
```
"kind" -> of[AssociateKind],
```

File: app/forms/AssociateUcr.scala (hunk 1 above)
Comment (bambuchaAdm) [reply app/forms/AssociateUcr.scala]:
```
error] /home/lukasz.dubiel/hmrc/customs-movements-frontend/app/forms/AssociateUcr.scala:50: Sealed trait forms.AssociateKind is not supported: no known subclasses
[error]   implicit val format = Json.format[AssociateKind]
[error]                                    ^
```

File: app/controllers/consolidations/AssociateDucrController.scala [hunk 6]
```diff
@@ -57,13 +57,15 @@ class AssociateDucrController @Inject()(
     form
       .bindFromRequest()
       .fold(
-        formWithErrors =>
+        formWithErrors => {
+          println(formWithErrors)
```
Comment (PatrykRudnicki) [inline app/controllers/consolidations/AssociateDucrController.scala]: Please remove println

File: app/forms/AssociateUcr.scala (hunk 1 above)
Comment (PatrykRudnicki) [reply app/forms/AssociateUcr.scala]: But now `AssociateKind` is a sealed trait and it should works (`Json.format`)

Comment (bambuchaAdm) [reply app/forms/AssociateUcr.scala]: I check it for now it does not.

File: app/views/associate_ucr.scala.html (hunk 4 above)
Comment (bambuchaAdm) [reply app/views/associate_ucr.scala.html]: Done - but probably require little enhancement later

File: app/views/associate_ucr.scala.html [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+
+@main_template(title = Title(messages("associate.ucr.title", mucrOptions.mucr))) {
+    @helper.form(consolidations.routes.AssociateDucrController.submit(), 'autoComplete -> "off") {
+        @helper.CSRF.formField
+
+        @components.back_link(consolidations.routes.MucrOptionsController.displayPage())
+
+        @components.error_summary(form.errors)
+
+        @components.page_title(Some(messages("associate.ucr.title", mucrOptions.mucr)))
```
Comment (MaciejRewera) [inline app/views/associate_ucr.scala.html]: Nit: In my opinion these 3 could live outside the form.

File: app/forms/AssociateUcr.scala [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+    }.getOrElse(Left(Seq(FormError(key, "error.required"))))
+
+    override def unbind(key: String, value: AssociateKind): Map[String, String] = Map(key -> value.formValue)
+  }
+
+  implicit val format =
+    Format[AssociateKind](Reads.StringReads.collect(JsonValidationError("error.unknown"))(lookup), Writes(kind => JsString(kind.formValue)))
+}
+
+case class AssociateUcr(kind: AssociateKind, ducr: Option[String], mucr: Option[String]) {
```
Comment (MaciejRewera) [inline app/forms/AssociateUcr.scala]: I'd suggest better names. Maybe `parentUcr` and `childUcr`?

File: test/views/AssociateDucrConfirmationViewSpec.scala [hunk 9]
```diff
@@ -73,6 +75,21 @@ class AssociateDucrConfirmationViewSpec extends UnitViewSpec with CommonMessages
       backButton.text() mustBe messages("site.backToStartPage")
       backButton.first() must haveHref(routes.ChoiceController.displayChoiceForm())
     }
+
+    "link to list of submissions" in {
+      pending
```
Comment (MaciejRewera) [inline test/views/AssociateDucrConfirmationViewSpec.scala]: 👍

File: test/views/DisassociateDucrConfirmationViewSpec.scala [hunk 10]
```diff
@@ -30,7 +30,7 @@ import views.tags.ViewTest
 class DisassociateDucrConfirmationViewSpec extends UnitViewSpec with CommonMessages {
 
   private val page = new disassociate_ducr_confirmation(mainTemplate)
-  private val view: Html = page()(request, new Flash(Map(FlashKeys.DUCR -> correctUcr)), messages)
+  private val view: Html = page()(request, new Flash(Map(FlashKeys.UCR -> correctUcr)), messages)
```
Comment (MaciejRewera) [inline test/views/DisassociateDucrConfirmationViewSpec.scala]: Isn't `CONSOLIDATION_KIND` required here?

File: test/views/MucrOptionsViewSpec.scala [hunk 11]
```diff
@@ -18,44 +18,44 @@ package views
...
-import play.twirl.api.Html
-import views.spec.ViewSpec
+import views.spec.UnitViewSpec
 
-class MucrOptionsViewSpec extends ViewSpec with CommonMessages {
+class MucrOptionsViewSpec extends UnitViewSpec with CommonMessages {
 
   private val form: Form[MucrOptions] = MucrOptions.form
-  private val page = injector.instanceOf[views.html.mucr_options]
+  private val page = new views.html.mucr_options(mainTemplate)
```
Comment (MaciejRewera) [inline test/views/MucrOptionsViewSpec.scala]: 👍

**Pull request merged**